src/internal/permissions.ts

```diff
@@ -0,0 +1,418 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+ * This file contains the underlying implementations for exposed API surface in
+ * the {@link api/name}. By moving the methods out into a separate file,
+ * other namespaces and processes can access these methods without depending on the entire
+ * name namespace and without having a dependency cycle error.
+ */
+
+import { AptosConfig } from "../api/aptosConfig";
+import { AbstractedEd25519Account, Account, SingleKeyAccount } from "../account";
+import { AccountAddress, AccountAddressInput } from "../core";
+import { SimpleTransaction } from "../transactions/instances/simpleTransaction";
+import { Aptos } from "../api";
+import { BatchArgument } from "@wgb5445/aptos-intent-npm";
+import { MoveString } from "../bcs";
+import { AptosIntentBuilder } from "../transactions";
+
+// types
+export type Permission = FungibleAssetPermission | GasPermission | NFTPermission | NFTCollectionPermission | GenericPermission;
+export type RevokePermission = RevokeFungibleAssetPermission | RevokeNFTPermission | Permission;
+export type FilteredPermissions<T extends PermissionType> = Array<
+  Extract<Permission, { type: T }>
+>;
+
+//  holds permissions
+export interface PermissionHandle {
+  // Question: Best way to approach time or date here. Date object, Epoch time string, time in milliseconds
+  // The contract will have this as an epoch time in seconds as a number
+  expiration: number;
+  // Question: Should this be a field at all? Should dapps be able to raise their own rate limit?
+  // Question: Should this be a number (1, 10, 100, etc) or a string (low, medium, high)?
+  transactionsPerSecond: number;
+  permissions: Permission[];
+}
+
+// Actions that can be granted on NFTs
+export enum NFTCapability {
+  transfer = "transfer",
+  mutate = "mutate",
+}
+
+export enum PermissionType {
+  FungibleAsset = "FungibleAsset",
+  Gas = "Gas",
+  NFT = "NFT",
+  NFTCollection = "Collection",
+}
+
+export enum MoveVMPermissionType {
+  FungibleAsset = "0x1::fungible_asset::WithdrawPermission",
+  
+}
+
+export interface FungibleAssetPermission {
+  type: PermissionType.FungibleAsset;
+  asset: string;
```

Comment:
Let's discuss about the two types of fungible permissions (primary store vs fixed amount)